### DISCUSSION TITLE
what is the meaning of iterations in dataloader?

### DISCUSSION THREAD

**Discussion by ppsmk388** [pyg-team/pytorch_geometric#5485] [Q&A] [answered]
I have found a word "iterations" in
https://pytorch-geometric.readthedocs.io/en/latest/modules/loader.html?highlight=HGTLoader#torch_geometric.loader.HGTLoader

It say “ Sample 512 nodes per type and per iteration for 4 iterations”，can you tell me what is the meaning of "iterations " in dataloader?

loader = HGTLoader(
    hetero_data,
    # Sample 512 nodes per type and per iteration for 4 iterations
    num_samples={key: [512] * 4 for key in hetero_data.node_types},
    # Use a batch size of 128 for sampling training nodes of type paper
    batch_size=128,
    input_nodes=('paper', hetero_data['paper'].train_mask),
)

**Comment by rusty1s** [comment, marked as answer]
The word iterations here mean how many steps are being used to sample neighbors. `HGTLoader` is a layer-wise sampling strategy, and so we will sample 512 nodes in 4 iterations/layers. Everytime we have sampled nodes in one layer, they will be added as seed nodes so that we can also sample their neighbors in the follow-up iteration.

**Reply by ppsmk388** [reply]
Thanks a lot！

**Comment by ppsmk388** [comment]
and for the code in https://github.com/pyg-team/pytorch_geometric/blob/master/examples/hetero/to_hetero_mag.py
I found this code:


train_loader = HGTLoader(**data**, num_samples=[1024] * 4, shuffle=True,
                         input_nodes=train_input_nodes, **kwargs)
val_loader = HGTLoader(**data**, num_samples=[1024] * 4,
                       input_nodes=val_input_nodes, **kwargs)


if the data from the training set and the data from the validation set all storage in the same graph :**data**

Does the data obtained from train_loader  contain the data in the validation set?

**Reply by EdisonLeeeee** [reply]
There shall be an overlap on the neighborhoods between the data obtained from the two loaders.

**Reply by ppsmk388** [reply]
thank you！

**Reply by ppsmk388** [reply]
but what is the meaning of the **train_mask**  in train_input_nodes = ('paper', data['paper'].train_mask)

**Reply by EdisonLeeeee** [reply]
The `train_mask` only specifies the root nodes (or seed nodes) in a sampled subgraph. For an inductive scenario, we only care about whether the root nodes overlap between training and test graphs. It's okay if there is an overlap between neighborhoods.

**Reply by ppsmk388** [reply]
thank you very much！！！

**Comment by ppsmk388** [comment]
Or  the data obtained from train_loader only contain the data from training set(beacuse of the train_mask and val_mask)?
I found the code below in https://github.com/pyg-team/pytorch_geometric/blob/master/examples/hetero/to_hetero_mag.py


**train_input_nodes** = ('paper', data['paper'].**train_mask**)
**val_input_nodes** = ('paper', data['paper'].**val_mask**)

train_loader = HGTLoader(data, num_samples=[1024] * 4, shuffle=True,
                         input_nodes=**train_input_nodes**, **kwargs)
val_loader = HGTLoader(data, num_samples=[1024] * 4,
                       input_nodes=**val_input_nodes**, **kwargs)
